blog/_posts/2017-11-27-view-based-collections.md

```diff
@@ -0,0 +1,163 @@
+Instead of having strict by default operations, they are now lazy by default and
+concrete collection types decide whether to evaluate their result or not. For instance,
+here is the implementation of `map`:
+
+~~~ scala
+trait IterableOps[A, CC[_], C] {
+  def map[B](f: A => B): CC[B] = fromIterable(View.Map(this, f))
+}
+~~~
+
+(`CC[_]` is the type constructor of the collection, e.g. `List` or `Vector`)
+
+We can break down this implementation in two steps.
+
+First, we create a lazy `View` that “records” that the `map` operation has been applied.
+Constructing such a view is a cheap operation, here is the implementation of
+`View.Map`:
+
+~~~ scala
+object View {
+  case class Map[A, B](underlying: Iterable[A], f: A => B) extends View[B] {
+    def iterator() = underlying.iterator().map(f)
+  }
+}
+~~~
+
+As you can see, unless we actually iterate on the view we don’t evaluate neither the elements
+of the `underlying` collection nor the mapped elements.
+
+Second, we build the resulting collection from the view by calling `fromIterable`. This operation
+is abstract in `IterableOps` and is implemented by leaf classes of the hierarchy (e.g.
+`List`, `Vector`, etc.). Here is the implementation of `fromIterable` in `List`,
+for instance:
+
+~~~ scala
+trait List[A] extends IterableOps[A, List, List[A]] {
+  def fromIterable[E](it: Iterable[E]): List[E] = (List.newBuilder[E]() ++= it).result()
+}
+~~~
+
+This implementation creates a new list and adds all the elements of the passed `Iterable[E]`
+to it. By doing so, it necessarily computes the elements of the source collection
+and the resulting collection: `List` is a strict collection type.
+
+Now let’s have a look at the `fromIterable` implementation in `View`:
+
+~~~ scala
+trait View[A] extends IterableOps[A, View, View[A]] {
+  def fromIterable[E](it: Iterable[E]): View[E] =
+    new View[E] {
+      def iterator() = it.iterator()
+    }
+}
+~~~
+
+Here we return a view that will only evaluate its elements (by delegating to the
+`underlying` collection elements) when it is actually iterated: `View` is a lazy
+collection type.
+
+Let’s take a step back and consider again the implementation of the `map` operation:
+
+~~~ scala
+trait IterableOps[A, CC[_], C] {
+  def map[B](f: A => B): CC[B] = fromIterable(View.Map(this, f))
+}
+~~~
+
+The first part of the implementation creates a lazy view of the transformed elements,
+and the second part of the implementation (the call to `fromIterable`) turns this view
+into the resulting collection (which can be done either in a lazy or strict way, as
+we’ve just seen).
+
+## Conclusion
+
+We have classified collections type in two categories: strict collections eagerly evaluate
+their elements (examples of such collections are `List`, `Vector`, etc.), whereas lazy
+collections evaluate their elements only when they are needed (examples are `Stream` and
+`View`).
+
+In the new collections, the default implementations of transformation operations preserve
```

Comment:
Would it make sense to add a comment about `withFilter` somewhere in the post? It seems `withFilter` should no longer be needed if `filter` is lazy by default.

Reply:
I think this part was unclear, I’ve rephrased it, let me know if you still have a question about `withFilter`.

Reply:
It's clearer now, thanks!